Panels/SharedStateDecorator.js

```diff
@@ -0,0 +1,150 @@
+import hoc from '@enact/core/hoc';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+const SharedState = React.createContext(null);
+
+const defaultConfig = {
+	idProp: 'id',
+	updateOnMount: false
+};
+
+/**
+ * Adds shared state to a component.
+ *
+ * The purpose of shared state is to store framework component state at significant container
+ * boundaries in order to restore it when the "same" component is mounted later.
+ *
+ * "Sameness" is determined by the `idProp` config member (defaults to "id"). If multiple
+ * descendants have the same `idProp` within the subtree, SharedStateDecorator will not distinguish
+ * between them and will allow each to read from and write over each other's data.
+ *
+ * For example, Panels and Panel are considered "significant container boundaries" since they are
+ * key building blocks for moonstone applications. When components are rendered within a Panel, we
+ * may want to store those components state on unmount so that we can restore it when returning to
+ * the panel. Panel can (and does) use SharedStateDecorator to establish a shared state which can be
+ * used by contained components.
+ *
+ * It's important to note that SharedStateDecorator doesn't prescribe how or what is stored nor how
+ * the data is managed. That is left to the consuming component to determine. Also, unlike React
+ * state or third-party state management solutions like Redux, updating shared state will not
+ * initiate an update cycle in React. The intent is only to restore state on mount.
+ *
+ * If shared state is used in the render method for a component, it may be necessary to use the
+ * `updateOnMount` config member which will initiate an update cycle within React once the data is
+ * available from an upstream shared state.
+ *
+ * @hoc
+ * @private
+ */
+const SharedStateDecorator = hoc(defaultConfig, (config, Wrapped) => {
```

Comment:
Like `IdProvider` but we might have a deeper team discussion about whether agate should provide this functionality.